include/af/opencv3.h

```diff
@@ -0,0 +1,191 @@
+
+    \returns an af::array with the contents of all cv::Mat in the std::vector
+*/
+static af::array array(const std::vector<cv::Mat>& input, bool transpose = true);
+
+/**
+   This function will copy data from a cv::Mat to an \ref af::array
+
+   \param[in]  input The data which will be loaded into the array
+   \param[in]  transpose flag specifying if the input should be transposed during copy
+
+   \returns an af::array with the contents of the OpenCv Mat
+*/
+static af::array array(const cv::Mat& input, bool transpose = true);
+
+/*
+ * arrayfire -> opencv
+ */
+/**
+   This function will copy data from an \ref af::array to a cv::Mat
+
+   \param[in]  input The af::array from which data which will be loaded into cv::Mat
+   \param[in]  type The resulting datatype of the output cv::Mat
+   \param[in]  transpose flag specifying if the input should be transposed during copy
+
+   \returns an cv::Mat with the contents of the input af::array
+*/
+static cv::Mat toMat(const af::array& input, int type = CV_32F, bool transpose = true);
+/**
+  @}
+*/
+
+
+//helper conversion function
+static void copy_data(cv::Mat& input, af::array& output, bool transpose) {
+    const unsigned w = input.cols;
+    const unsigned h = input.rows;
+    const unsigned nchannels = input.channels();
+    if (nchannels > 4) { throw af::exception("mat to array error "); }
+    if (nchannels == 1) {
+        // bw, guaranteed to be continuous
+        if (transpose) {
+            output = af::array(w, h, input.ptr<float>(0)).T();
+        } else {
+            output = af::array(w, h, input.ptr<float>(0));
+        }
+    } else {
+        if(input.isContinuous()) {
+            std::vector<cv::Mat> channels; split(input, channels);
+            size_t dims = channels.size();
+
+            if (transpose) {
+                output = af::array(h, w, dims);
+                for(unsigned d=0; d<dims; ++d) {
+                    output(af::span, af::span, d) = af::array(w, h, channels[d].ptr<float>(0)).T();
+                }
+            } else {
+                output = af::array(w, h, dims);
+                for(unsigned d=0; d<dims; ++d) {
+                    output(af::span, af::span, d) = af::array(w, h, channels[d].ptr<float>(0));
+                }
+            }
+        } else {
+            throw af::exception("mat to array error ");
+        }
+    }
+}
+
+
+/*
+ * opencv -> arrayfire
+ */
+
+// conversion for cv::Mat
+static af::array array(const cv::Mat& input, bool transpose)
+{
+    af::array output;
+    if (input.empty()) { return output; }
+    cv::Mat tmp;
+    if (input.channels() == 1)
+        { input.convertTo(tmp, CV_32F); }
+    else if (input.channels() == 2)
+        { input.convertTo(tmp, CV_32FC2); }
+    else if (input.channels() == 3)
+        { input.convertTo(tmp, CV_32FC3); }
+    else if (input.channels() == 4)
+        { input.convertTo(tmp, CV_32FC4); }
+
+    copy_data(tmp, output, transpose);
```

Comment:
There are several things wrong here.

- copy_data is only used once in this file.
- It should also not be used by the user because it performs no checks.
- The function is not actually performing a copy. It is creating a new output array.

Just move all the relevant code here.